### ISSUE TITLE
strip content from pull request body on merge

### ISSUE THREAD

**Issue by clxmstaab** [chdsbd/kodiak#550]
we are using https://pullrequestbadge.com/ which adds pull request badges into our PR descriptions based on keywords in the PR title (e.g. a JIRA Issue key)

example how this looks like:

![image](https://user-images.githubusercontent.com/47448731/97979732-f8878980-1dcf-11eb-93f4-a91fce18f9c2.png)

the badges get inserted by a github marketplace app automatically.

since we use the PR-body as a commit message, commits contains this ugly markdown:

![image](https://user-images.githubusercontent.com/47448731/97979819-1d7bfc80-1dd0-11eb-8c78-77bfc259c882.png)

we would love to sorround the generated markup with e.g. a html comment, so kodiak could strip it when merging.

what do you think?

we have the following kodiak config regarding merge.

```
[merge.message]
title = "pull_request_title"
body = "pull_request_body"
include_pr_number = true
strip_html_comments = true # remove html comments to auto remove PR templates
```

**Comment by clxmstaab** [comment]
@stefanbuck would it be possible to sorround the generated badges with some "marker" markup, so tools like kodiak could strip it?

or maybe even pullrequestbadge itself could strip the markup again before the PR gets merged?

**Comment by chdsbd** [comment]
I think if the content was wrapped with HTML with a specific `id` the content would be easy to identify and remove.

```
<div id="pull-request-badge">
[![CircleCI](https://circleci.com/gh/chdsbd/kodiak.svg?style=svg&circle-token=4879604a0cca6fa815c4d22936350f5bdf455905)](https://circleci.com/gh/chdsbd/kodiak)
</div>
```

```toml
# .kodiak.toml

[merge.message]
strip_html_ids = ["pull-request-badge"] # remove HTML nodes with `id`

```


We could also maybe do something with comments and remove content between specific comments.
```
<!-- pull-request-badge:start -->
[![CircleCI](https://circleci.com/gh/chdsbd/kodiak.svg?style=svg&circle-token=4879604a0cca6fa815c4d22936350f5bdf455905)](https://circleci.com/gh/chdsbd/kodiak)
<!-- pull-request-badge:end -->
```

```toml
# .kodiak.toml

[merge.message]
strip_html_comment_spans = ["pull-request-badge"]
```


I think the challenge is finding syntax that looks okay if it's _not_ removed.

**Comment by stefanbuck** [comment]
> or maybe even pullrequestbadge itself could strip the markup again before the PR gets merged?

Possible, but that way you remove badges from merged PRs. Personally I find it quite useful when looking at an older PR (particular  the JIRA badge). However, there is already a separator comment: `<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->`. Everything before this comment can safely removed since Pull Request Badge insert badges on top always.

**Comment by sbdchd** [comment]
I think to be safe, it would be better if there was a closing comment after the PR Badge content.
That way we don't accidentally delete someone's commit message if they happen to put stuff at the top.

So I think a generic solution would look something like:

```
text before badges

<!-- PR-BADGE: whatever text you want -->
badge content here
<!-- PR-BADGE: can put any text here as well -->

text after badges
```

and then Kodiak would delete any content inside comments that begin with a prefix like `PR-BADGE`

I think the config @chdsbd outlined would work with this comment setup:

```toml
# .kodiak.toml

[merge.message]
strip_html_comment_spans = ["PR-BADGE"]
```

**Comment by stefanbuck** [comment]
I see your point, but this is not possible. Pull Request Badge runs whenever the description is added or updated. That way badges are always on top. If you still have strong feeling I'm happy to add such an opening comment

**Comment by sbdchd** [comment]
Gotcha, I think if there is an opening comment as well, then this feature can be used with any tool that adds stuff to a PR body.

With only the closing comment we'd have to do a one-off for PR Badge which is doable, but it would be nice if we added something more general for other bots

**Comment by ardeois** [comment]
I would be interested in a similar feature, but I would like to suggest a different configuration approach.

At my work, we used [bors](https://bors.tech/) in the past but switch to Kodiak later, as it's more flexible. Bors has a [configuration](https://bors.tech/documentation/#configuration-borstoml) named `cut_body_after` :
> A marker in the PR description that indicates boilerplate that does not belong in the commit message.

We found this config very useful, and would like a similar feature for Kodiak.

What would you think if we had:
```toml
# .kodiak.toml
[merge.message]
cut_body_before = "kodiak-message:start"
cut_body_after = "kodiak-message:end"
```

The marker could even be an html comment so it's not seen in the PR description:
```toml
# .kodiak.toml
[merge.message]
cut_body_before = "<!-- kodiak-message:start -->"
cut_body_after = "<!-- kodiak-message:end -->"
```

This way we would be in full control of the commit message and that should cover your use case @clxmstaab ?

What do you think?

**Comment by chdsbd** [comment]
@ardeois's suggestion seems like it would work.
```toml
# .kodiak.toml
[merge.message]
cut_body_before = "<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->"
```

I think we'd need to add a couple config options to: https://github.com/chdsbd/kodiak/blob/59d155d90c07952e56174ad2e1a3f59851694cdd/bot/kodiak/config.py#L33-L45

And update this function to use those new config options: https://github.com/chdsbd/kodiak/blob/59d155d90c07952e56174ad2e1a3f59851694cdd/bot/kodiak/evaluation.py#L65-L77

**Comment by stefanbuck** [comment]
@chdsbd please let me know if you still need the opening comment. I just had no time to do it, but should be easy.

**Comment by chdsbd** [comment]
@stefanbuck I think we're all set know thanks to @ardeois's suggestion.

With the "cut_body_before" option we can remove the badges by specifying the PR-Badge comment.

**Comment by clxmstaab** [comment]
As I understood it, the solution of @ardeois will work for us, in case @stefanbuck will add the necessary markup

**Comment by chdsbd** [comment]
[`merge.message.cut_body_before`](https://kodiakhq.com/docs/config-reference#mergemessagecut_body_before) and [`merge.message.cut_body_after`](https://kodiakhq.com/docs/config-reference#mergemessagecut_body_after) are now released via #589 and #595. 


So to remove the PR Badge labels from a message like:

```
[![CircleCI](https://circleci.com/gh/chdsbd/kodiak.svg?style=svg&circle-token=4879604a0cca6fa815c4d22936350f5bdf455905)](https://circleci.com/gh/chdsbd/kodiak)
<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

Add new feature "foo" to accept user feedback.

```

We can use the following config:

```toml
# .kodiak.toml
[merge.message]
body = "pull_request_body"
cut_body_before = "<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->"
strip_html_comments = true
```

And on merge we'll have the following commit message:

```

Add new feature "foo" to accept user feedback.

```

The `merge.message.cut_body_before` option will remove the content for the badges and `merge.message.strip_html_comments` will remove the "PR-BADGE: PLEASE DO NOT REMOVE" comment.

Please open an issue if you run into any problems.

**Comment by clxmstaab** [comment]
just stumbled upon another use-case.

when merging dependabot-PRs using kodiak we get a long list of useless stuff at the end of the merge-commit.

example:

![image](https://user-images.githubusercontent.com/47448731/102195310-b4070780-3ebe-11eb-8bd3-ddd5b3b8cf29.png)


we will use `cut_body_before ="Dependabot will resolve any conflicts with this PR as long as you don't alter it yourself"` as a mean to make this shorter. lets see if it works.
